components/google-drive/changes-to-any-file-in-drive.js

```diff
@@ -0,0 +1,203 @@
+// This source processes changes to any files in a user's Google Drive,
+// implementing strategy enumerated in the Push Notifications API docs:
+// https://developers.google.com/drive/api/v3/push and here:
+// https://developers.google.com/drive/api/v3/manage-changes
+//
+// This source has two interfaces:
+//
+// 1) The HTTP requests tied to changes in the user's Google Drive
+// 2) A timer that runs on regular intervals, renewing the notification channel as needed
+
+const { uuid } = require("uuidv4");
+const includes = require("lodash.includes");
+const googleDrive = require("https://github.com/PipedreamHQ/pipedream/blob/google-drive-changes-to-file/components/google-drive/google-drive.app.js");
+
+module.exports = {
+  name: "Google Drive - Changes to Any Files in My Drive",
+  version: "0.0.1",
+  // Dedupe events based on the "x-goog-message-number" header for the target channel:
+  // https://developers.google.com/drive/api/v3/push#making-watch-requests
+  dedupe: "unique",
+  props: {
+    googleDrive,
+    db: "$.service.db",
+    http: "$.interface.http",
+    updateTypes: { propDefinition: [googleDrive, "updateTypes"] },
+    watchForPropertiesChanges: {
+      propDefinition: [googleDrive, "watchForPropertiesChanges"],
+    },
+    timer: {
+      label: "Push notification renewal schedule",
+      description:
+        "The Google Drive API requires occasionaly renewal of push notification subscriptions. **This runs in the background, so you should not need to modify this schedule**.",
+      type: "$.interface.timer",
+      default: {
+        intervalSeconds: 60 * 60 * 6,
+      },
+    },
+  },
+  hooks: {
+    async activate() {
+      // Called when a componenent is created or updated. Handles all the logic
+      // for starting and stopping watch notifications tied to the desired files.
+
+      const channelID = this.db.get("channelID") || uuid();
+
+      // Subscriptions are tied to Google's resourceID, "an opaque value that
+      // identifies the watched resource". This value is included in request headers
+      let subscription = this.db.get("subscription");
+
+      const startPageToken = await this.googleDrive.getPageToken();
+      const { expiration, resourceId } = await this.googleDrive.watchDrive(
+        channelID,
+        this.http.endpoint,
+        startPageToken
+      );
+      // We use and increment the pageToken as new changes arrive, in run()
+      this.db.set("pageToken", startPageToken);
+
+      // Save metadata on the subscription so we can stop / renew later
+      this.db.set("subscription", { resourceId, expiration });
+      this.db.set("channelID", channelID);
+    },
+    async deactivate() {
+      const channelID = this.db.get("channelID");
```

Comment:
maybe so that db is cleared in all cases?

```
const channelID = ...
const { resourceId } = ...
if (channelID && resourceId) {
  await this.googleDrive...
}
this.db.set("subscription", null);
this.db.set("channelID", null);
// also clear pageToken ?
```

also why the ID vs Id ?
```

Reply:
`go lint` fucked me: https://github.com/golang/lint/issues/124

Reply:
I use `resourceId` because that's the name of the param Google accepts, but otherwise I use ID when I control it. I should probably change to be consistent

Reply:
lol